src/librustc/ty/constness.rs

```diff
@@ -63,13 +64,63 @@ impl<'tcx> TyCtxt<'tcx> {
 
 
 pub fn provide(providers: &mut Providers<'_>) {
-    /// only checks whether the function has a `const` modifier
+    /// Const evaluability whitelist is here to check evaluability at the
+    /// top level beforehand.
+    fn is_const_intrinsic(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+        match tcx.fn_sig(def_id).abi() {
+            Abi::RustIntrinsic |
+            Abi::PlatformIntrinsic => {
+                match tcx.item_name(def_id) {
+                    | sym::size_of
+                    | sym::min_align_of
+                    | sym::needs_drop
+                    | sym::type_id
+                    | sym::bswap
+                    | sym::bitreverse
+                    | sym::ctpop
+                    | sym::cttz
+                    | sym::cttz_nonzero
+                    | sym::ctlz
+                    | sym::ctlz_nonzero
+                    | sym::overflowing_add
+                    | sym::overflowing_sub
+                    | sym::overflowing_mul
+                    | sym::unchecked_shl
+                    | sym::unchecked_shr
+                    | sym::rotate_left
+                    | sym::rotate_right
+                    | sym::add_with_overflow
+                    | sym::sub_with_overflow
+                    | sym::mul_with_overflow
+                    | sym::saturating_add
+                    | sym::saturating_sub
+                    | sym::transmute
+                    | sym::wrapping_add
+                    | sym::wrapping_sub
+                    | sym::wrapping_mul
+                    | sym::simd_extract
+                    | sym::simd_insert
+                    | sym::caller_location
+                    | sym::type_name
+                    => true,
+
+                    _ => false
+                }
+            }
+            _ => false
```

Comment:
Shouldn't this `bug!` as it is not an intrinsic?

Reply:
No, the function is called before we know it's an intrinsic

Reply:
That seems like an odd choice IMO.

Reply:
I can make it return an option. But I see this function as `is_intrinsic(did) && is_const_intrinsic(did)`

Reply:
Adding a clarifying doc comment would be enough. :)